core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -766,49 +813,73 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
+                                     } else {
+                                        filterToUse = filter;
+                                     }
+                                     return storeStatus.<K, Object>store().publishKeys(segments, filterToUse);
+                                  }
+                               }
+                               return Flowable.empty();
+                            },
+                            this::releaseReadLock);
    }
 
    @Override
    public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, boolean localInvocation,
-         boolean includeStores) {
+                                                                            boolean includeStores) {
       return loadFromAllStores(key, keyPartitioner.getSegment(key), localInvocation, includeStores);
    }
 
    @Override
-   public <K, V> CompletionStage<MarshallableEntry<K, V>> loadFromAllStores(Object key, int segment,
-         boolean localInvocation, boolean includeStores) {
-      return Maybe.using(
-            this::acquireReadLock,
-            ignore -> {
-               checkStoreAvailability();
-               if (log.isTraceEnabled()) {
-                  log.tracef("Loading entry for key %s with segment %d", key, segment);
-               }
-               return Flowable.fromIterable(stores)
-                     .filter(storeStatus -> allowLoad(storeStatus, localInvocation, includeStores))
-                     // Only do 1 request at a time
-                     .concatMapMaybe(storeStatus -> Maybe.fromCompletionStage(
-                           storeStatus.<K, V>store().load(segmentOrZero(storeStatus, segment), key)), 1)
-                     .firstElement();
-            },
-            this::releaseReadLock
-      ).toCompletionStage(null);
+   public <K, V> CompletionStage<MarshallableEntry<K, V>>
+   loadFromAllStores(Object key, int segment, boolean localInvocation, boolean includeStores) {
+      long stamp = acquireReadLock();
+      boolean release = true;
+      try {
+         checkStoreAvailability();
+         if (log.isTraceEnabled()) {
+            log.tracef("Loading entry for key %s with segment %d", key, segment);
+         }
+         Iterator<StoreStatus> iterator = stores.iterator();
+         CompletionStage<MarshallableEntry<K, V>> stage = loadFromStoresIterator(key, segment, iterator, localInvocation, includeStores);
+         if (CompletionStages.isCompletedSuccessfully(stage)) {
+            return stage;
+         } else {
+            release = false;
+            return stage.whenComplete((e, throwable) -> releaseReadLock(stamp));
+         }
+      } finally {
+         if (release) {
+            releaseReadLock(stamp);
+         }
+      }
+   }
+
+   private <K, V> CompletionStage<MarshallableEntry<K, V>>
+   loadFromStoresIterator(Object key, int segment, Iterator<StoreStatus> iterator, boolean localInvocation, boolean includeStores) {
```

Comment:
is it intentional the method name be on a new line?

Reply:
Yes, I'm trying to keep lines under 120 characters and align parameter names when splitting lines, and long return types don't make it easy :)